clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -129,21 +130,29 @@ public void clearAssignment() {
     }
 
     public Long fetched(TopicPartition tp) {
-        return this.fetched.get(tp);
+        return assertAssignment(tp).fetched;
     }
 
     public void fetched(TopicPartition tp, long offset) {
-        if (!this.assignedPartitions.contains(tp))
-            throw new IllegalArgumentException("Can't change the fetch position for a partition you are not currently subscribed to.");
-        this.fetched.put(tp, offset);
+        if (!isFetchable(tp))
+            // we only allow updates to the fetched position when both fetched and consumed positions are valid
+            throw new IllegalStateException("Cannot update fetch position of a partition not in a fetchable state ");
+        assertAssignment(tp).fetched(offset);
+    }
+
+    private TopicPartitionState assertAssignment(TopicPartition tp) {
```

Comment:
Can we rename it to something like "getState", assertPartition is kind of misleading.